src/mono/mono/metadata/class.c

```diff
@@ -1379,6 +1379,129 @@ mono_method_set_generic_container (MonoMethod *method, MonoGenericContainer* con
 	mono_image_property_insert (mono_method_get_image (method), method, MONO_METHOD_PROP_GENERIC_CONTAINER, container);
 }
 
+/**
+ * mono_method_set_verification_success:
+ *
+ * Sets a bit indicating that the method has been verified.
+ *
+ * LOCKING: acquires the image lock.
+ */
+void
+mono_method_set_verification_success (MonoMethod *method)
+{
+	g_assert (!method->is_inflated);
+
+	mono_image_property_insert (mono_method_get_image (method), method, MONO_METHOD_PROP_VERIFICATION_SUCCESS, GUINT_TO_POINTER(1));
+}
+
+/**
+ * mono_method_get_verification_sucess:
+ *
+ * Returns \c TRUE if the method has been verified successfully.
+ *
+ * LOCKING: acquires the image lock.
+ */
+gboolean
+mono_method_get_verification_success (MonoMethod *method)
+{
+	if (method->is_inflated)
+		method = ((MonoMethodInflated *)method)->declaring;
+
+	gpointer value = mono_image_property_lookup (mono_method_get_image (method), method, MONO_METHOD_PROP_VERIFICATION_SUCCESS);
+
+	return value != NULL;
+}
+
+/**
+ * mono_method_lookup_infrequent_bits:
+ *
+ * Looks for existing \c MonoMethodDefInfrequentBits struct associated with
+ * this method definition.
+ *
+ * LOCKING: Acquires the image lock
+ */
+const MonoMethodDefInfrequentBits*
+mono_method_lookup_infrequent_bits (MonoMethod *method)
+{
+	g_assert (!method->is_inflated);
+
+	return (const MonoMethodDefInfrequentBits*)mono_image_property_lookup (mono_method_get_image (method), method, MONO_METHOD_PROP_INFREQUENT_BITS);
+}
+
+/**
+ * mono_method_get_infrequent_bits:
+ *
+ * Looks for an existing, or allocates a new \c MonoMethodDefInfrequentBits struct for this method definition.
+ * Method must not be inflated.
+ *
+ * LOCKING: Acquires the image lock
+ */
+MonoMethodDefInfrequentBits *
+mono_method_get_infrequent_bits (MonoMethod *method)
+{
+	g_assert (!method->is_inflated);
+	MonoImage *image = mono_method_get_image (method);
+	MonoMethodDefInfrequentBits *infrequent_bits = NULL;
+	mono_image_lock (image);
+	infrequent_bits = (MonoMethodDefInfrequentBits *)mono_image_property_lookup (image, method, MONO_METHOD_PROP_INFREQUENT_BITS);
+	if (!infrequent_bits) {
+		infrequent_bits = (MonoMethodDefInfrequentBits *)mono_image_alloc0 (image, sizeof (MonoMethodDefInfrequentBits));
+		mono_image_property_insert (image, method, MONO_METHOD_PROP_INFREQUENT_BITS, infrequent_bits);
```

Comment:
I assume the image lock is recursive and this is fine?

Reply:
yes, it's recursive. I checked when I was writing this :D